raft/src/main/java/org/apache/kafka/raft/internals/SerdeRecordsIterator.java

```diff
@@ -0,0 +1,231 @@
+                final ByteBuffer buffer;
+                if (allocatedBuffer.isPresent()) {
+                    buffer = allocatedBuffer.get();
+                    buffer.compact();
+
+                    if (!buffer.hasRemaining()) {
+                        // The buffer is not big enough to read an entire batch
+                        throw new IllegalStateException(
+                            String.format(
+                                "Unable to read batch from file records buffer %s with maximum batch %s and record size %s",
+                                buffer,
+                                maxBatchSize,
+                                records.sizeInBytes()
+                            )
+                        );
+                    }
+                } else {
+                    buffer = bufferSupplier.get(Math.min(maxBatchSize, records.sizeInBytes()));
+                    allocatedBuffer = Optional.of(buffer);
+                }
+
+                int start = buffer.position();
+                try { 
+                    ((FileRecords) records).readInto(buffer, bytesRead);
+                } catch (IOException e) {
+                    throw new RuntimeException("Failed to read records into memory", e);
+                }
+
+                bytesRead += buffer.limit() - start;
+                memoryRecords = MemoryRecords.readableRecords(buffer.slice());
+            } else {
+                throw new IllegalStateException(String.format("Unexpected Records type %s", records.getClass()));
+            }
+
+            return Optional.of(memoryRecords.batchIterator());
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private Optional<Batch<T>> nextBatch() {
+        if (!nextBatches.isPresent()) {
+            nextBatches = nextBatches();
+        }
+
+        while (nextBatches.isPresent()) {
+            if (nextBatches.get().hasNext()) {
+                MutableRecordBatch nextBatch = nextBatches.get().next();
+
+                // Update the buffer position to reflect the read batch
+                allocatedBuffer.ifPresent(buffer -> buffer.position(buffer.position() + nextBatch.sizeInBytes()));
+
+                if (!(nextBatch instanceof DefaultRecordBatch)) {
+                    throw new IllegalStateException(
+                        String.format("DefaultRecordBatch expected by record type was %s", nextBatch.getClass())
+                    );
+                }
+                return Optional.of(readBatch((DefaultRecordBatch) nextBatch));
+            }
+
+            nextBatches = nextBatches();
+        }
+
+        return Optional.empty();
+    }
+
+    private Batch<T> readBatch(DefaultRecordBatch batch) {
+        final Batch<T> result;
+        if (batch.isControlBatch()) {
+            result = Batch.empty(batch.baseOffset(), batch.partitionLeaderEpoch(), batch.lastOffset());
+        } else {
+            Integer numRecords = batch.countOrNull();
```

Comment:
Maybe use OptionalInt here? We're pretty consistently using Optional everywhere else in this class

Reply:
Probably a good move. I think this dates back before we had `Optional` available. I'd suggest doing it in a separate PR though since it's going to touch a bunch of otherwise unrelated code.

Reply:
Yes. I can do this in a followup PR. Do we want to add a new method or do we want to replace the existing method? Is `interface RecordBatch` considered a public API?

Reply:
Separate PR sounds fine. I'm pretty sure `RecordBatch` is a public interface though. I'm not sure how we normally go about changing signatures in public APIs. Definitely out of scope for this PR